File: agentic_pipeline/pipelineTypes.ts
// LAYER 1: pipeline types + structural assertions (TypeScript port of pipeline_types.py).
//
// Two things live here, same as the Python:
//   1. interfaces mirroring the JSON schemas (documentation + typing).
//   2. runtime assert* functions called at each stage boundary. these fail fast
//      with a precise message when a stage produces the wrong shape.
//
// The assert* functions are the runnable part. they take plain `any` (the Python
// took plain dicts and used .get()), so they can check possibly-malformed data.
//
// NOTE on naming: pipeline_types.py called the beat-level {method, params, offset}
// shape `VizAction`, but that name means something different in schemas/types.ts
// (the plan-level {method, description, params_schema}). to avoid the collision we
// call the beat-level one `BeatVizAction` here. the plan-level shape keeps the name
// `VizActionSpec` (same as in the Python). the interfaces are type-only (erased at
// runtime), so the overlap with schemas/types.ts costs nothing.

// ---- primitive building blocks ----

export type ActId = string; // ^act_[a-z0-9_]+$
export type GateId = string; // ^gate_[a-z0-9_]+$
export type VizModeLiteral = "svg" | "figure" | "chart" | null;
export type GateType = "quiz" | "fill-in" | "proof-builder" | "interactive";
export type NodeType = "act" | "gate" | "marker";

export interface BeatVizAction {
  method: string;
  params?: Record<string, any>;
  offset?: string | number;
}

export interface Beat {
  say?: string;
  card?: Record<string, any> | null;
  viz_actions?: BeatVizAction[];
  inline_viz?: string | boolean | null;
  duration?: number | null;
  viz_panel_override?: string | null;
}

export interface ActSpec {
  act_id: ActId;
  title: string;
  viz_panel: "svg" | "figure" | "chart" | null;
  beats: Beat[];
}

export interface BlankConfig {
  answer?: string[];
  width?: number;
  placeholder?: string;
}

export interface GateSpec {
  gate_id?: GateId; // injected by orchestrator from plan node id
  gate_type?: GateType;
  after_act?: ActId;
  question?: string;
  options?: string[];
  correct?: number | string;
  explanations?: Record<string, string>;
  prompt?: string;
  blank?: BlankConfig;
  hint?: string;
  successMessage?: string;
  instruction?: string;
  availablePieces?: Record<string, string>[];
  correctOrder?: string[];
  slots?: number;
  title?: string;
  description?: string;
  slider?: Record<string, any>;
  compute?: string;
  displays?: Record<string, string>[];
  challenge?: string;
  wrong_path_acts?: ActId[];
}

export interface VizSpec {
  mode?: "preset" | "custom_code" | "mobject_plugin" | "three_js";
  preset?: string | null;
  config?: Record<string, any> | null;
  code?: string | null;
  mobject_plugin_code?: string | null;
  three_js_code?: string | null;
  actions_implemented?: string[];
}

export interface BeatOutline {
  narration_hint?: string;
  card_type?: string | null;
  viz_actions?: string[];
  inline_at_end?: boolean;
}

export interface ActNode {
  type?: "act";
  id?: ActId;
  title?: string;
  objective?: string;
  viz_panel?: "svg" | "figure" | "chart" | null;
  beat_outline?: BeatOutline[];
  context_from_previous?: string;
  wrong_path_acts?: ActNode[];
}

export interface GateNode {
  type?: "gate";
  id?: GateId;
  gate_type?: GateType;
  after_act?: ActId;
  question_hint?: string;
  wrong_path_hint?: string;
  wrong_path_acts?: ActNode[];
}

export interface MarkerNode {
  type: "marker";
  label: string;
  after_act: ActId;
}

export type PlanNode = ActNode | GateNode | MarkerNode;

export interface VizActionSpec {
  method?: string;
  description?: string;
  params_schema?: Record<string, string>;
}

export interface VizRequirements {
  type?: "custom" | "preset_number_line" | "preset_coord_plane" | "preset_unit_circle" | "none";
  description?: string;
  actions?: VizActionSpec[];
  config?: Record<string, any>;
  viewBox?: string;
}

export interface PlanMeta {
  title?: string;
  source?: string;
  answer?: string;
  estimated_duration_minutes?: number;
}

export interface PlanProblem {
  text?: string;
  highlight?: string;
}

export interface LessonPlan {
  meta: PlanMeta;
  problem: PlanProblem;
  viz_requirements: VizRequirements;
  nodes: PlanNode[];
}

export type ActSpecMap = Record<string, ActSpec>;
export type GateSpecMap = Record<string, GateSpec>;


// ─────────────────────────────────────────────────────────────────────
// Runtime structural assertions, called at every stage boundary.
// Each throws TypeError with a "[pipeline contract] ..." message naming the bad
// field. they do NOT mutate their argument.
// ─────────────────────────────────────────────────────────────────────

const ACT_ID_RE = /^act_[a-z0-9_]+$/;
const GATE_ID_RE = /^gate_[a-z0-9_]+$/;
const VALID_GATE_TYPES = new Set(["quiz", "fill-in", "proof-builder", "interactive"]);
const VALID_NODE_TYPES = new Set(["act", "gate", "marker"]);
const VALID_VIZ_MODES = new Set(["preset", "custom_code", "mobject_plugin", "three_js"]);

// True for a plain object (Python isinstance(x, dict)), excluding arrays/null.
function isDict(x: any): boolean {
  return x !== null && typeof x === "object" && !Array.isArray(x);
}

// Rough match for Python's type(x).__name__, used only in error text.
function typeName(x: any): string {
  if (x === null || x === undefined) return "NoneType";
  if (Array.isArray(x)) return "list";
  if (typeof x === "object") return "dict";
  if (typeof x === "string") return "str";
  if (typeof x === "number") return Number.isInteger(x) ? "int" : "float";
  if (typeof x === "boolean") return "bool";
  return typeof x;
}

// Rough match for Python's {x!r}: quote strings, None for null/undefined.
function pyRepr(x: any): string {
  if (x === null || x === undefined) return "None";
  if (typeof x === "string") return `'${x}'`;
  return String(x);
}

// Throw unless cond is truthy. single failure mode for every assert_* check.
function require_(cond: any, msg: string): void {
  if (!cond) {
    throw new TypeError(`[pipeline contract] ${msg}`);
  }
}

// Return [firstIndex, dupIndex] of the first repeated value, or null.
function findDuplicate(xs: any[]): [number, number] | null {
  // null-proto map so prototype-name keys ("toString", "valueOf", "__proto__", ...)
  // behave like a Python dict — no inherited members to false-match against.
  const seen: Record<string, number> = Object.create(null);
  for (let i = 0; i < xs.length; i++) {
    const x = xs[i];
    if (x in seen) return [seen[x], i];
    seen[x] = i;
  }
  return null;
}

// Throw unless `plan` matches LessonPlan. Called after stage 2.
export function assertPlanShape(plan: any): void {
  require_(isDict(plan), `plan must be dict, got ${typeName(plan)}`);
  require_("meta" in plan && "title" in plan.meta, "plan.meta.title missing");
  require_("problem" in plan && "text" in plan.problem, "plan.problem.text missing");
  require_(Array.isArray(plan.nodes), "plan.nodes must be list");

  const ids = new Set<string>();
  plan.nodes.forEach((n: any, i: number) => {
    require_(isDict(n), `plan.nodes[${i}] not a dict`);
    const t = n.type;
    require_(VALID_NODE_TYPES.has(t), `plan.nodes[${i}].type=${pyRepr(t)} invalid`);
    if (t === "act" || t === "gate") {
      const nid = n.id;
      require_(typeof nid === "string", `plan.nodes[${i}].id missing`);
      if (t === "act") {
        require_(ACT_ID_RE.test(nid), `act id ${pyRepr(nid)} must match ^act_[a-z0-9_]+$`);
      } else {
        require_(GATE_ID_RE.test(nid), `gate id ${pyRepr(nid)} must match ^gate_[a-z0-9_]+$`);
        require_(VALID_GATE_TYPES.has(n.gate_type), `gate ${nid} has invalid gate_type ${pyRepr(n.gate_type)}`);
        require_(typeof n.after_act === "string", `gate ${nid}.after_act missing`);
      }
      require_(!ids.has(nid), `duplicate node id: ${nid}`);
      ids.add(nid);
    }
  });
}

// Throw unless `spec` matches ActSpec. With plan_node, also checks act_id match,
// beat count match, and unique beat 'say' fields.
export function assertActSpecShape(spec: any, planNode: any = null): void {
  require_(isDict(spec), `act spec not a dict: ${typeName(spec)}`);
  require_(ACT_ID_RE.test(spec.act_id ?? ""), `act_id missing/invalid: ${pyRepr(spec.act_id)}`);
  const beats = spec.beats;
  require_(Array.isArray(beats) && beats.length > 0, `act ${spec.act_id}: beats must be non-empty list`);
  beats.forEach((b: any, i: number) => {
    require_(isDict(b), `act ${spec.act_id} beat ${i} not dict`);
    require_(typeof b.say === "string" && b.say.trim(), `act ${spec.act_id} beat ${i} missing/empty 'say'`);
  });

  const says = beats.map((b: any) => b.say);
  const dupIdx = findDuplicate(says);
  if (dupIdx !== null) {
    throw new TypeError(
      `[pipeline contract] act ${spec.act_id}: duplicate beat text at ` +
        `positions (${dupIdx[0]}, ${dupIdx[1]}). Act worker produced repeated narration. ` +
        `Offending text: ${pyRepr(String(says[dupIdx[0]]).slice(0, 100))}`
    );
  }

  if (planNode !== null) {
    require_(spec.act_id === planNode.id, `act_id ${pyRepr(spec.act_id)} != plan node id ${pyRepr(planNode.id)}`);
    const outline = planNode.beat_outline || [];
    if (outline.length) {
      require_(
        beats.length === outline.length,
        `act ${spec.act_id}: produced ${beats.length} beats, plan ` +
          `beat_outline has ${outline.length}. Act worker dropped or ` +
          `duplicated beats.`
      );
    }
  }
}

// Throw unless `spec` matches GateSpec after orchestrator injection (gate_id present
// and authoritative). With plan_node, checks gate_id/gate_type/after_act all match.
export function assertGateSpecShape(spec: any, planNode: any = null): void {
  require_(isDict(spec), `gate spec not a dict: ${typeName(spec)}`);
  require_(GATE_ID_RE.test(spec.gate_id ?? ""), `gate_id missing/invalid (must be injected from plan): ${pyRepr(spec.gate_id)}`);
  require_(VALID_GATE_TYPES.has(spec.gate_type), `gate ${spec.gate_id}: invalid gate_type ${pyRepr(spec.gate_type)}`);
  require_(
    typeof spec.after_act === "string" && ACT_ID_RE.test(spec.after_act),
    `gate ${spec.gate_id}: after_act missing/invalid`
  );

  if (planNode !== null) {
    require_(
      spec.gate_id === planNode.id,
      `gate_id mismatch: spec=${pyRepr(spec.gate_id)} plan=${pyRepr(planNode.id)}. ` + `Assembler lookup would fail.`
    );
    require_(spec.gate_type === planNode.gate_type, `gate ${spec.gate_id}: gate_type mismatch with plan`);
    require_(spec.after_act === planNode.after_act, `gate ${spec.gate_id}: after_act mismatch with plan`);
  }
}

// Throw unless `spec` matches VizSpec for its declared mode.
export function assertVizSpecShape(spec: any): void {
  require_(isDict(spec), "viz spec not dict");
  require_(VALID_VIZ_MODES.has(spec.mode), `viz spec mode ${pyRepr(spec.mode)} invalid`);
  if (spec.mode === "custom_code") {
    require_(typeof spec.code === "string" && spec.code.trim(), "custom_code mode requires non-empty 'code'");
  } else if (spec.mode === "mobject_plugin") {
    require_(
      typeof spec.mobject_plugin_code === "string" && spec.mobject_plugin_code.trim(),
      "mobject_plugin mode requires non-empty 'mobject_plugin_code'"
    );
  } else if (spec.mode === "three_js") {
    require_(
      typeof spec.three_js_code === "string" && spec.three_js_code.trim(),
      "three_js mode requires non-empty 'three_js_code'"
    );
  } else if (spec.mode === "preset") {
    require_(typeof spec.preset === "string" && spec.preset.trim(), "preset mode requires non-empty 'preset'");
  }
}

// Throw unless every action declared in plan.viz_requirements.actions appears in
// viz_spec.actions_implemented. catches silent drops by the viz agent.
export function assertVizImplementsPlanActions(vizSpec: any, plan: any): void {
  const declared = new Set<string>(
    (plan.viz_requirements?.actions || []).map((a: any) => a.method)
  );
  const implemented = new Set<string>(vizSpec.actions_implemented || []);
  const missing = [...declared].filter((m) => !implemented.has(m));
  if (missing.length) {
    missing.sort();
    // match Python's repr of a sorted list: ['m2', 'm3']
    const reprList = `[${missing.map((m) => `'${m}'`).join(", ")}]`;
    throw new TypeError(
      `[pipeline contract] viz_spec is missing ${missing.length} action(s) ` +
        `declared in the plan: ${reprList}. Viz agent silently dropped them.`
    );
  }
}

File: agentic_pipeline/python_originals/pipeline_types.py
"""
Pipeline artifact type definitions.

All TypedDicts mirror the JSON schemas in schemas/. Use these for:
  - mypy static checking across the pipeline
  - documentation of what each stage produces/consumes
  - runtime assertions at stage boundaries

NOTE: gate_id is absent from GateSpec — the orchestrator injects it from the
plan node's "id" field. The LLM never produces gate_id.
"""

from __future__ import annotations
from typing import Any, Dict, List, Literal, Optional, Union
try:
    from typing import TypedDict
except ImportError:
    from typing_extensions import TypedDict


# ─────────────────────────────────────────────────────────────────────
# Primitive building blocks
# ─────────────────────────────────────────────────────────────────────

ActId   = str  # must match ^act_[a-z0-9_]+$
GateId  = str  # must match ^gate_[a-z0-9_]+$
VizMode = Literal["svg", "figure", "chart", None]
GateType = Literal["quiz", "fill-in", "proof-builder", "interactive"]
NodeType = Literal["act", "gate", "marker"]


class VizAction(TypedDict, total=False):
    method: str                          # required
    params: Dict[str, Any]              # optional, defaults to {}
    offset: Union[str, int, float]       # optional, defaults to 0


class Beat(TypedDict, total=False):
    say: str                             # required — narration text
    card: Optional[Dict[str, Any]]       # optional — card definition
    viz_actions: List[VizAction]         # optional, defaults to []
    inline_viz: Optional[Union[str, bool]]  # "svg" | "figure" | "chart" | True
    duration: Optional[float]            # override auto-estimated duration
    viz_panel_override: Optional[str]    # per-beat vizPanel mode override


# ─────────────────────────────────────────────────────────────────────
# Stage 3a output: ActSpec
# ─────────────────────────────────────────────────────────────────────

class ActSpec(TypedDict):
    act_id: ActId
    title: str
    viz_panel: Optional[Literal["svg", "figure", "chart"]]
    beats: List[Beat]


# ─────────────────────────────────────────────────────────────────────
# Stage 3b output: GateSpec
#
# gate_id is NOT part of LLM output. The orchestrator injects it:
#   spec["gate_id"] = node["id"]   # always from the plan, never from LLM
# ─────────────────────────────────────────────────────────────────────

class BlankConfig(TypedDict, total=False):
    answer: List[str]       # required — accepted answers
    width: int              # optional — input width in pixels
    placeholder: str        # optional — placeholder text


class GateSpec(TypedDict, total=False):
    # Injected by pipeline — not LLM output:
    gate_id: GateId          # set by orchestrator from plan node["id"]

    # LLM produces these:
    gate_type: GateType      # required
    after_act: ActId         # required — must match a plan act node id

    # Quiz fields:
    question: str
    options: List[str]
    correct: Union[int, str]
    explanations: Dict[str, str]

    # Fill-in fields:
    prompt: str
    blank: BlankConfig
    hint: str
    successMessage: str

    # Proof-builder fields:
    instruction: str
    availablePieces: List[Dict[str, str]]  # [{id, latex}]
    correctOrder: List[str]
    slots: int

    # Interactive fields:
    title: str
    description: str
    slider: Dict[str, Any]
    compute: str
    displays: List[Dict[str, str]]
    challenge: str

    # Wrong path:
    wrong_path_acts: List[ActId]


# ─────────────────────────────────────────────────────────────────────
# Stage 4 output: VizSpec
# ─────────────────────────────────────────────────────────────────────

class VizSpec(TypedDict, total=False):
    mode: Literal["preset", "custom_code", "mobject_plugin", "three_js"]  # required
    preset: Optional[str]             # if mode == "preset"
    config: Optional[Dict[str, Any]]  # passed to L.viz()
    code: Optional[str]               # if mode == "custom_code"
    mobject_plugin_code: Optional[str]  # if mode == "mobject_plugin"
    three_js_code: Optional[str]      # if mode == "three_js"
    actions_implemented: List[str]    # method names the plugin implements


# ─────────────────────────────────────────────────────────────────────
# Stage 2 output: LessonPlan nodes
# ─────────────────────────────────────────────────────────────────────

class BeatOutline(TypedDict, total=False):
    narration_hint: str           # required
    card_type: Optional[str]
    viz_actions: List[str]        # method names (not yet parameterized)
    inline_at_end: bool


class ActNode(TypedDict, total=False):
    type: Literal["act"]          # required
    id: ActId                     # required — unique, ^act_[a-z0-9_]+$
    title: str                    # required
    objective: str                # required
    viz_panel: Optional[Literal["svg", "figure", "chart"]]  # required (can be null)
    beat_outline: List[BeatOutline]  # required
    context_from_previous: str
    wrong_path_acts: List["ActNode"]  # for branch acts


class GateNode(TypedDict, total=False):
    type: Literal["gate"]         # required
    id: GateId                    # required — unique, ^gate_[a-z0-9_]+$
    gate_type: GateType           # required
    after_act: ActId              # required — must reference an existing act id
    question_hint: str
    wrong_path_hint: str
    wrong_path_acts: List[ActNode]  # branch acts defined inline


class MarkerNode(TypedDict):
    type: Literal["marker"]
    label: str
    after_act: ActId


PlanNode = Union[ActNode, GateNode, MarkerNode]


class VizActionSpec(TypedDict, total=False):
    method: str          # required
    description: str     # required
    params_schema: Dict[str, str]  # param_name -> type string


class VizRequirements(TypedDict, total=False):
    type: Literal["custom", "preset_number_line", "preset_coord_plane",
                  "preset_unit_circle", "none"]  # required
    description: str     # required
    actions: List[VizActionSpec]  # required — the action contract
    config: Dict[str, Any]
    viewBox: str


class PlanMeta(TypedDict, total=False):
    title: str           # required
    source: str          # required
    answer: str
    estimated_duration_minutes: float


class PlanProblem(TypedDict, total=False):
    text: str            # required
    highlight: str


class LessonPlan(TypedDict):
    meta: PlanMeta
    problem: PlanProblem
    viz_requirements: VizRequirements
    nodes: List[PlanNode]


# ─────────────────────────────────────────────────────────────────────
# Pipeline-level types
# ─────────────────────────────────────────────────────────────────────

ActSpecMap  = Dict[ActId, ActSpec]
GateSpecMap = Dict[GateId, GateSpec]


# ─────────────────────────────────────────────────────────────────────
# Runtime structural assertions — called at every stage boundary.
#
# 6.031 "avoiding debugging" applied: fail fast with a precise error that
# names the stage, the field, and the expected shape. These are cheap
# guards that catch contract violations before they silently propagate
# through subsequent stages (the way the gate-drop bug did).
#
# Each asserter raises TypeError with a message that pinpoints the bad
# input. They do NOT mutate their argument.
# ─────────────────────────────────────────────────────────────────────

import re as _re

_ACT_ID_RE  = _re.compile(r"^act_[a-z0-9_]+$")
_GATE_ID_RE = _re.compile(r"^gate_[a-z0-9_]+$")
_VALID_GATE_TYPES = {"quiz", "fill-in", "proof-builder", "interactive"}
_VALID_NODE_TYPES = {"act", "gate", "marker"}
_VALID_VIZ_MODES  = {"preset", "custom_code", "mobject_plugin", "three_js"}


def _require(cond, msg):
    """Raise TypeError(`[pipeline contract] {msg}`) unless `cond` is truthy.

    @param cond: predicate — True to proceed silently, False to raise.
    @param msg: short explanation shown to the developer (not the end user).
    @raises TypeError: always, whenever cond is falsy. Used as the single failure
        mode for every `assert_*_shape` check so callers can catch one exception type.
    """
    if not cond:
        raise TypeError(f"[pipeline contract] {msg}")


def assert_plan_shape(plan):
    """Raise TypeError unless `plan` matches LessonPlan. Called after stage 2."""
    _require(isinstance(plan, dict), f"plan must be dict, got {type(plan).__name__}")
    _require("meta" in plan and "title" in plan["meta"], "plan.meta.title missing")
    _require("problem" in plan and "text" in plan["problem"], "plan.problem.text missing")
    _require(isinstance(plan.get("nodes"), list), "plan.nodes must be list")

    ids = set()
    for i, n in enumerate(plan["nodes"]):
        _require(isinstance(n, dict), f"plan.nodes[{i}] not a dict")
        t = n.get("type")
        _require(t in _VALID_NODE_TYPES, f"plan.nodes[{i}].type={t!r} invalid")
        if t in ("act", "gate"):
            nid = n.get("id")
            _require(isinstance(nid, str), f"plan.nodes[{i}].id missing")
            if t == "act":
                _require(_ACT_ID_RE.match(nid),
                         f"act id {nid!r} must match ^act_[a-z0-9_]+$")
            else:
                _require(_GATE_ID_RE.match(nid),
                         f"gate id {nid!r} must match ^gate_[a-z0-9_]+$")
                _require(n.get("gate_type") in _VALID_GATE_TYPES,
                         f"gate {nid} has invalid gate_type {n.get('gate_type')!r}")
                _require(isinstance(n.get("after_act"), str),
                         f"gate {nid}.after_act missing")
            _require(nid not in ids, f"duplicate node id: {nid}")
            ids.add(nid)


def assert_act_spec_shape(spec, plan_node=None):
    """Raise unless `spec` matches ActSpec. If plan_node provided, also checks:
       - spec.act_id == plan_node.id
       - len(spec.beats) == len(plan_node.beat_outline)  (no duplicates, no drops)
       - beat 'say' fields are unique (catches the duplicate-beat bug).
    """
    _require(isinstance(spec, dict), f"act spec not a dict: {type(spec).__name__}")
    _require(_ACT_ID_RE.match(spec.get("act_id", "")),
             f"act_id missing/invalid: {spec.get('act_id')!r}")
    beats = spec.get("beats")
    _require(isinstance(beats, list) and len(beats) > 0,
             f"act {spec.get('act_id')}: beats must be non-empty list")
    for i, b in enumerate(beats):
        _require(isinstance(b, dict), f"act {spec['act_id']} beat {i} not dict")
        _require(isinstance(b.get("say"), str) and b["say"].strip(),
                 f"act {spec['act_id']} beat {i} missing/empty 'say'")

    says = [b["say"] for b in beats]
    dup_idx = _find_duplicate(says)
    if dup_idx is not None:
        raise TypeError(
            f"[pipeline contract] act {spec['act_id']}: duplicate beat text at "
            f"positions {dup_idx}. Act worker produced repeated narration. "
            f"Offending text: {says[dup_idx[0]][:100]!r}"
        )

    if plan_node is not None:
        _require(spec["act_id"] == plan_node["id"],
                 f"act_id {spec['act_id']!r} != plan node id {plan_node['id']!r}")
        outline = plan_node.get("beat_outline", [])
        if outline:
            _require(len(beats) == len(outline),
                     f"act {spec['act_id']}: produced {len(beats)} beats, plan "
                     f"beat_outline has {len(outline)}. Act worker dropped or "
                     f"duplicated beats.")


def _find_duplicate(xs):
    """Return (first_index, duplicate_index) of the first repeated value in `xs`, or None.

    @param xs: iterable of hashable values (typically beat 'say' strings).
    @returns: (i, j) with i < j and xs[i] == xs[j], or None if every element is unique.
    """
    seen = {}
    for i, x in enumerate(xs):
        if x in seen:
            return (seen[x], i)
        seen[x] = i
    return None


def assert_gate_spec_shape(spec, plan_node=None):
    """Raise unless `spec` matches GateSpec after orchestrator injection.

    Post-injection the spec MUST contain gate_id (authoritative, from plan),
    after_act, and gate_type. This is what the assembler looks up; mismatches
    here were the critical gate-drop bug.
    """
    _require(isinstance(spec, dict), f"gate spec not a dict: {type(spec).__name__}")
    _require(_GATE_ID_RE.match(spec.get("gate_id", "")),
             f"gate_id missing/invalid (must be injected from plan): {spec.get('gate_id')!r}")
    _require(spec.get("gate_type") in _VALID_GATE_TYPES,
             f"gate {spec.get('gate_id')}: invalid gate_type {spec.get('gate_type')!r}")
    _require(isinstance(spec.get("after_act"), str) and _ACT_ID_RE.match(spec["after_act"]),
             f"gate {spec.get('gate_id')}: after_act missing/invalid")

    if plan_node is not None:
        _require(spec["gate_id"] == plan_node["id"],
                 f"gate_id mismatch: spec={spec['gate_id']!r} plan={plan_node['id']!r}. "
                 f"Assembler lookup would fail.")
        _require(spec["gate_type"] == plan_node["gate_type"],
                 f"gate {spec['gate_id']}: gate_type mismatch with plan")
        _require(spec["after_act"] == plan_node["after_act"],
                 f"gate {spec['gate_id']}: after_act mismatch with plan")


def assert_viz_spec_shape(spec):
    """Raise TypeError unless `spec` matches VizSpec for its declared mode.

    @param spec: viz_spec dict emitted by stage 3 (or loaded from viz_spec.json).
    @raises TypeError: if `mode` is unknown, or if the mode-specific code field
        (code / mobject_plugin_code / three_js_code / preset) is missing or empty.
    """
    _require(isinstance(spec, dict), "viz spec not dict")
    _require(spec.get("mode") in _VALID_VIZ_MODES,
             f"viz spec mode {spec.get('mode')!r} invalid")
    if spec["mode"] == "custom_code":
        _require(isinstance(spec.get("code"), str) and spec["code"].strip(),
                 "custom_code mode requires non-empty 'code'")
    elif spec["mode"] == "mobject_plugin":
        _require(isinstance(spec.get("mobject_plugin_code"), str) and spec["mobject_plugin_code"].strip(),
                 "mobject_plugin mode requires non-empty 'mobject_plugin_code'")
    elif spec["mode"] == "three_js":
        _require(isinstance(spec.get("three_js_code"), str) and spec["three_js_code"].strip(),
                 "three_js mode requires non-empty 'three_js_code'")
    elif spec["mode"] == "preset":
        _require(isinstance(spec.get("preset"), str) and spec["preset"].strip(),
                 "preset mode requires non-empty 'preset'")


def assert_viz_implements_plan_actions(viz_spec, plan):
    """Every action declared in plan.viz_requirements.actions must appear in
       viz_spec.actions_implemented. Catches silent drops by the viz agent."""
    declared = {a["method"] for a in plan.get("viz_requirements", {}).get("actions", [])}
    implemented = set(viz_spec.get("actions_implemented", []))
    missing = declared - implemented
    if missing:
        raise TypeError(
            f"[pipeline contract] viz_spec is missing {len(missing)} action(s) "
            f"declared in the plan: {sorted(missing)}. Viz agent silently dropped them."
        )
